src/celeritas/optical/interactor/WavelengthShiftInteractor.hh

```diff
@@ -0,0 +1,174 @@
+ * optical property of wavelength shifters. The polarization of the reemitted
+ * lights is assumed to be incoherent with respect to the polarization of the
+ * primary optical photon.
+ *
+ * \note This performs the same sampling routine as in the G4OpWLS class of
+ * the Geant4 release 11.2.
+ */
+class WavelengthShiftInteractor
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Energy = units::MevEnergy;
+    using ParamsRef = NativeCRef<WavelengthShiftData>;
+    using SecondaryAllocator = StackAllocator<TrackInitializer>;
+    //!@}
+
+  public:
+    // Construct with shared and state data
+    inline CELER_FUNCTION
+    WavelengthShiftInteractor(ParamsRef const& shared,
+                              ParticleTrackView const& particle,
+                              OpticalMaterialId const& mat_id,
+                              SecondaryAllocator& allocate);
+
+    // Sample an interaction with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng);
+
+  private:
+    //// DATA ////
+
+    // Incident photon energy
+    Energy const inc_energy_;
+    // Sample distributions
+    PoissonDistribution<real_type> sample_num_photons_;
+    ExponentialDistribution<real_type> sample_time_;
+    // Grid calculators
+    GenericCalculator calc_energy_;
+    GenericCalculator calc_cdf_;
+    // Allocate space for secondary particles
+    SecondaryAllocator& allocate_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ */
+CELER_FUNCTION
+WavelengthShiftInteractor::WavelengthShiftInteractor(
+    ParamsRef const& shared,
+    ParticleTrackView const& particle,
+    OpticalMaterialId const& mat_id,
+    SecondaryAllocator& allocate)
+    : inc_energy_(particle.energy())
+    , sample_num_photons_(shared.wls_record[mat_id].mean_num_photons)
+    , sample_time_(real_type{1} / shared.wls_record[mat_id].time_constant)
+    , calc_energy_(shared.energy_cdf[mat_id], shared.reals)
+    , calc_cdf_(GenericCalculator::from_inverse(shared.energy_cdf[mat_id],
+                                                shared.reals))
+    , allocate_(allocate)
+{
+    CELER_EXPECT(inc_energy_.value() > calc_energy_(0));
+}
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Sampling the wavelength shift (WLS) photons as in G4OpWLS of the Geant4
+ * release 11.2 with a modification in sampling the reemission spectrum.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction WavelengthShiftInteractor::operator()(Engine& rng)
+{
+    // Sample the number of photons generated from WLS
+    unsigned int num_photons = sample_num_photons_(rng);
```

Comment:
Network access is so slow here - almost impossible to do anything. Can you help to add the note? Otherwise, I will do it when I get back  (this weekend). Thanks.

Reply:
No problem!

Reply:
I've annotated the to-do.